Curriculum learning rework

dericp opened the pull request:
The curriculum learning system now supports multiple curriculums for multiple brains. Each brain can have a corresponding curriculum.

The way curriculums are specified is now different. We start by defining a "metacurriculum" which is basically a dictionary from brain_name to Curriculum. A MetaCurriculum object replaces the Curriculum object in `trainer_controller.py`.

- Curriculum tests have been pulled out into a separate file.
- `meta_curriculum.py` module added under `unitytrainers/`.
- Curriculum and MetaCurriculum are unit tested.
- Curriculum learning documentation has been updated.
- Confirmed that multi-brain curriculums are working in Wall Jump environment.

awjuliani commented on python/unitytrainers/trainer.py:
```diff
@@ -150,7 +150,8 @@ def write_summary(self, lesson_number):
                     summary.value.add(tag='Info/{}'.format(key), simple_value=stat_mean)
                     self.stats[key] = []
-            summary.value.add(tag='Info/Lesson', simple_value=lesson_number)
+            # summary.value.add(tag='Info/Lesson', simple_value=lesson_number)
```
Commented line.

awjuliani commented on python/unitytrainers/trainer_controller.py:
```diff
@@ -236,16 +236,16 @@ def start_learning(self):
-                        self.curriculum.increment_lesson(self._get_progress())
-                        curr_info = self.env.reset(config=self.curriculum.get_config(), train_mode=self.fast_simulation)
+                        self.meta_curriculum.increment_lessons(self._get_progresses())
+                        curr_info = self.env.reset(config=self.meta_curriculum.get_config(), train_mode=self.fast_simulation)
```
Wrap these lines too.

dericp replied:
Will refactor `trainer_controller.py` to comply to 80 character line limit in another PR.

awjuliani commented:
@dericp Overall looks good. My comments were mostly cosmetic. As Vince mentioned offline, we will want to increment lessons manually without waiting for the academy to reset. However, that can be a separate PR after this one.

dericp commented on docs/Training-Curriculum-Learning.md:
```diff
@@ -2,68 +2,97 @@
+
+We first create a folder inside `python/curricula/` for the environment we want
+to use curriculum learning with. For example, if we were creating a metacurriculum
+for Wall Jump, we would create the folder `python/curricula/wall-jump/`. We will place
```
I only bolded the first mention of Wall Jump so the reader can orient themselves as to what environment the rest of the document is using.

dericp commented:
Can I merge this or should I wait for more approvals? @awjuliani

vincentpierre commented on docs/Training-Curriculum-Learning.md:
```diff
@@ -2,68 +2,97 @@
+parameters of the environment will vary. In the case of the Wall Jump environment, what
+varies is the height of the wall. We define this as a `Reset Parameter` in the Academy
+object of our scene, and by doing so it becomes adjustable via the Python API. Rather
+than adjusting it by hand, we will create a simple JSON file which describes the
```
remove simple
It is not obvious that the json files must have the names of the brains in the scene. This should be important.

vincentpierre commented on docs/Training-Curriculum-Learning.md:
```diff
@@ -2,68 +2,97 @@
+Once our curriculum is defined, we have to use the reset parameters we defined and modify
+the environment from the agent's `AgentReset()` function. See
+[WallJumpAgent.cs](https://github.com/Unity-Technologies/ml-agents/blob/master/unity-environment/Assets/ML-Agents/Examples/WallJump/Scripts/WallJumpAgent.cs)
+for an example.
```
We should specify that if the academy does not reset, the reset parameters will not be updated

vincentpierre commented on python/unitytrainers/meta_curriculum.py:
```diff
@@ -0,0 +1,85 @@
+            self.brains_to_curriculums[brain_name].increment_lesson(progress)
+
+
+    def set_all_curriculums_to_lesson_num(self, lesson_num):
```
Do we want to be able to set all lessons to the same number or do we want to be able to set the lessons individually when using learn.py ?

dericp replied:
Since `lesson_nums` is a property it's easy to set all the lesson numbers individually. This function just makes it convenient for initializing all the lessons to a specific one.

vincentpierre commented on python/unitytrainers/meta_curriculum.py:
```diff
@@ -0,0 +1,85 @@
+            curriculum.lesson_num = lesson_num
+
+
+    def get_config(self):
```
I see how this can be complicated if multiple curriculums need to access the same reset parameters. Maybe in addition to display a warning, there should be an ordering of the curriculums : The lowest lesson numbers have priority or something like that.

dericp replied:
I'm not sure if that last part would be true. The ordering of the brain_name keys in the dict should be random so we really don't know which curriculum would take precedence.

vincentpierre replied:
What I meant is that we can we can scan through the keys of the dictionary intelligently rather than using the default dictionary order. We __make__ the precedence happen.

dericp replied:
I'm inclined to avoid that because of the inefficiency. However, I do see the utility of adding this---it makes the lowest lesson number, the one that needs the most help, the one that controls the parameter. Let me see if I can find a way to do this without the inefficiency.

vincentpierre commented on python/unitytrainers/curriculum.py:
```diff
@@ -73,17 +72,17 @@ def increment_lesson(self, progress):
                 for key in parameters:
-                    config[key] = parameters[key][self.lesson_number]
+                    config[key] = parameters[key][self.lesson_num]
                 logger.info("\nLesson changed. Now in Lesson {0} : \t{1}"
```
Since there are now multiple curricula, The message should display the name of the brain that is associated. I even feel all messages should be handled by the meta_curriculum.

dericp commented:
@vincentpierre Regarding the logging, I believe that log was there before I started working on the code. I agree that there is probably a better way to log information...let me try some stuff.

dericp commented:
@vincentpierre I think I've addressed your comments. The latest commit fixes the TensorBoard logging. I rebased and pushed so all the commits are under the original comments (sorry!).

vincentpierre commented on python/unitytrainers/trainer_controller.py:
```diff
@@ -86,21 +91,35 @@ def __init__(self, env_path, run_id, save_freq, curriculum_file, fast_simulation
+                    raise MetaCurriculumError('One of the curriculums '
+                                              'defined in ' +
+                                              self.curriculum_folder +
+                                              'does not have a corresponding '
```
space before does

dericp commented:
@vincentpierre I like your suggestion to make `NotADirectoryError` into a `MetaCurriculumError`. I've added this.

I think raising the error of the reset parameter first is okay. This will at least point the user in the direction that something is wrong with their metacurriculum setup. If they go into the file, they should quickly realize that they're passing the wrong metacurriculum. And, if they don't realize this, once they change the reset parameters a `MetaCurriculumError` will be raised telling them that the brain names are mismatched. Also, I think it would be better to avoid passing environment information into `MetaCurriculum` and `Curriculum`. It seems that it could be better to have all the environment manipulation and checking in `trainer_controller.py` so that `MetaCurriculum` and `Curriculum` remain as independent as possible.

vincentpierre commented on python/unitytrainers/meta_curriculum.py:
```diff
@@ -25,22 +25,29 @@ def __init__(self, curriculum_folder, default_reset_parameters):
                 if any([(parameter in curriculum.get_config().keys()) for parameter in used_reset_parameters]):
-                    logger.info('Two of more curriculums have attempted to change '
-                                'the same reset parameter. The result will be '
+                    logger.info('WARNING: Two of more curriculums will '
```
What about using `logger.warning` ?

dericp replied:
Good point. Changed.
